internal/server/logbuffer/merge.go

```diff
@@ -0,0 +1,146 @@
+
+	// Value returns the log entry itself. This allows time addition
+	// wrappers to return their original value.
+	Value() interface{}
+}
+
+// MergeReader is value that returns TimedEntry's for LogMerge
+// to weave together.
+type MergeReader interface {
+	Next() (TimedEntry, error)
+}
+
+// NewMerger creates a new LogMerge, with the stream generated
+// from the given inputs.
+func NewMerger(readers ...MergeReader) *Merger {
+	var lm Merger
+	lm.readers = readers
+
+	lm.heads = lm.makeHeads()
+
+	return &lm
+}
+
+// TimedEntries is a convience type of TimedEntry's that provides
+// the LogMergeInput interface.
+type TimedEntries []TimedEntry
+
+// Next returns the next value in the slice and then shrinks itself.
+func (t *TimedEntries) Next() (TimedEntry, error) {
+	if len(*t) == 0 {
+		return nil, io.EOF
+	}
+
+	ent := (*t)[0]
+
+	*t = (*t)[1:]
+
+	return ent, nil
+}
+
+// Create a slice to be used by refillEntries and findNext
+func (l *Merger) makeHeads() []TimedEntry {
+	return make([]TimedEntry, len(l.readers))
+}
+
+// Populate any missing entries with an entry from the corresponding
+// input (index in slice corresponds to inputs slice).
+func (l *Merger) refillEntries(entries []TimedEntry) int {
+	var pop int
+
+	for i, ent := range entries {
+		if ent != nil {
+			pop++
+			continue
+		}
+
+		ent, err := l.readers[i].Next()
+		if err == nil {
+			pop++
+			entries[i] = ent
+		}
+	}
+
+	return pop
+}
+
+// Find the entry in entries with earliest time, returning the entry and the
+// input that generated it.
+func (l *Merger) findNext(entries []TimedEntry) (TimedEntry, MergeReader) {
+	var (
+		best      TimedEntry
+		bestIdx   int
+		bestInput MergeReader
+	)
+
+	for i, ent := range entries {
+		if ent == nil {
+			continue
+		}
+
+		if best == nil || ent.Time().Before(best.Time()) {
+			best = ent
+			bestIdx = i
+			bestInput = l.readers[i]
+		}
+	}
+
+	if best == nil {
+		return nil, nil
+	}
+
+	entries[bestIdx] = nil
+
+	return best, bestInput
+}
+
+// ReaderEntry is returned by ReadNext. It provides access to the TimedEntry
+// that is next as well as the input that generated the entry. This
+// type is important because it allows the caller to figure out the context
+// of the entry from the input. Because LogMerge is going to effectively
+// shuffle the values that are put into it, the caller is going to have to deal
+// with entries appearing in any order and the input provides critical context.
+type ReaderEntry struct {
+	TimedEntry
+	Reader MergeReader
+}
+
+// ReadNext returns a slice of InputEntrys that are next in total time order.
+// The result might be fewer than count values, depending on what is available.
+func (l *Merger) ReadNext(count int) ([]ReaderEntry, error) {
+	var out []ReaderEntry
+
+	heads := l.heads
+
+	for i := 0; i < count; i++ {
+		pop := l.refillEntries(heads)
+		if pop == 0 {
+			break
+		}
+
+		ent, ip := l.findNext(heads)
+		out = append(out, ReaderEntry{TimedEntry: ent, Reader: ip})
+	}
```

Comment:
Hmmm, have we thought about moving the sorting to the constructor to do it upfront and then save on cost during the reads?

Reply:
Which sorting? This is the core of the algorithm and it has to happen here.

Reply:
Sorry no current sorting. I meant not having to do the lookup for earliest Entries each time you call Next() and instead move it to be done upfront once

Reply:
Because there isn't the ability to know to move it, that's how the algorithm works. There isn't visibility into which one will be earliest until all readers have had a chance to contribute a value.